### ISSUE TITLE
Injecting Custom Template Loading Logic

### ISSUE THREAD

**Issue by lekhmanrus** [angular/angular-cli#28600]
### Command

build, serve

### Description

In previous Angular versions, when using webpack-based builders, we had the ability to customize how component templates were processed (e.g., using third-party loaders like [ngx-pug-builders](https://github.com/lekhmanrus/ngx-pug-builders) to support Pug templates). With ESBuild, template loading logic to be managed internally by the `AdapterResourceLoader`.
It looks, like currently there appears to be no way to inject custom logic into the template loading process. This limits the ability for third-party libraries to provide support for alternative template formats or to apply custom modifications during the build.

### Describe the solution you'd like

It would be beneficial for Angular to expose a mechanism for developers to hook into or modify the HTML loading logic. This would enable third-party libraries to support template engines like Pug or apply other transformations.

While Angular CLI may not want to handle all possible template engines directly, allowing developers to customize the HTML loading process would open the door for greater flexibility and extensibility within the ecosystem.

**Comment by clydin** [comment]
Can you provide additional information regarding your use cases that are not met by using Angular templates?
While the Angular template language can be parsed as valid HTML, its full syntactic space goes beyond HTML. This is especially true with the control flow, @defer, @let, etc. capabilities now present. Additionally, the developer ecosystem tools rely on Angular components using the Angular template language and includes schematics, updates, linting (eslint), formatting (prettier), IDE integration (Angular language service or WebStorm), and more.

**Comment by lekhmanrus** [comment]
Hi @clydin,
I understand that integrating and supporting pug files isn't part of the Angular team's current plans, but a significant number of developers still use Pug in their projects. I had previously integrated the webpack pug loader via a custom Angular builders, but since Angular switched to esbuild and vite (which btw works perfectly and fast), integrating pug into Angular has become impossible. This seems to be because templates are loaded through `AdapterResourceLoader` (as I understand), and any esbuild plugins are ignored at that stage.
I also tried importing pug files using a loader kinda:
```ts
import template from './app.component.pug' with { loader: 'text' };
```
However, using this in the `@Component`'s `template` is also not possible.
Regarding control flow, `defer`, `let`, and similar capabilities, it's worth noting that they can still be used in pug templates as well. So perhaps it would be good to have some ability to inject into the template loading (to compile pug to html in this specific case)? Or maybe there is some better workaround.

P.S.: Personally I do not use Pug in my Angular projects anymore - just trying to support a library that some developers still seem to use.

**Comment by Eugeny** [comment]
To be clear, it's not about replacing the Angular templating functionality with Pug's, but rather using pure, non-templating Pug syntax with Angular, for example:

```html
<div class="dropdown-menu-right" ngbDropdownMenu>
   <a class="dropdown-item" (click)="setFilterMode(&quot;fuzzy&quot;)">
   <i class="far fa-fw fa-tilde"></i>
   <span translate>Fuzzy</span></a>
</div>
```

becomes 

```pug
.dropdown-menu-right(ngbDropdownMenu)
    a.dropdown-item((click)='setFilterMode("fuzzy")')
        i.far.fa-fw.fa-tilde
        span(translate) Fuzzy
```

**Comment by raykin** [comment]
Btw, I only use basic features of Pug to combine with Angular template. That's only use shortcut syntax of Pug and use all logic flow from Angular template. 
It help me get a clean code, write less and read fast and debug easier. 
So I don't care the logic control feature of Pug or other high level feature,  can we have alternatives? It looks like not a complex task, ex: convert . to wrapped div tags, convert a to wrapped a tags.
Sometimes I think maybe I can extract a new lib from Pug to do only basic converting tasks. Unfortunately I'm not a good js lib coder.
Another reason why I prefer such syntax is Emacs is my editor and I don't use plugins to auto convert html tags. I know it's not a complex to use html plugin, I just comfortable with the css style html code. Write less not only just good for write, but also good for read and debug.

**Comment by Eugeny** [comment]
So... I just had to roll it myself, and you can too with this `patch-package` patch:

`patches/@angular+compiler+17.3.11.patch`:

```diff
diff --git a/node_modules/@angular/compiler/fesm2022/compiler.mjs b/node_modules/@angular/compiler/fesm2022/compiler.mjs
index 521c187..778aa67 100755
--- a/node_modules/@angular/compiler/fesm2022/compiler.mjs
+++ b/node_modules/@angular/compiler/fesm2022/compiler.mjs
@@ -30263,6 +30263,9 @@ function getTextInterpolationExpression(interpolation) {
             return Identifiers.textInterpolateV;
     }
 }
+
+import { compile as compilePug } from 'pug'
+
 /**
  * Parse a template into render3 `Node`s and additional metadata, with no other dependencies.
  *
@@ -30271,6 +30274,10 @@ function getTextInterpolationExpression(interpolation) {
  * @param options options to modify how the template is parsed
  */
 function parseTemplate(template, templateUrl, options = {}) {
+    if (templateUrl.endsWith('.pug')) {
+        console.log('Compiling', templateUrl)
+        template = compilePug(template, { doctype: 'html', pretty: true })()
+    }
     const { interpolationConfig, preserveWhitespaces, enableI18nLegacyMessageIdFormat } = options;
     const bindingParser = makeBindingParser(interpolationConfig);
     const htmlParser = new HtmlParser();

```

**Comment by reduckted** [comment]
I would love this feature too. Using the custom webpack builder, I'm currently using a webpack plugin to inline SVG files into templates (SVGs for things like company logos, etc). This allows the SVG files to be used in multiple places without having to duplicate them in each template.

**Comment by iblislin** [comment]
Thanks @Eugeny. You're a lifesaver.

I ran into another issue while trying tests with Vitest, seems the `TestBed` will put a dummy value for `templateUrl`, e.g. `ng:///MyComponent2/template.html`. so I just add an extra condition to it:

```typescript
if (templateUrl.endsWith('.pug') || templateUrl.endsWith('/template.html')) {
...
}
```

**Comment by mihilion** [comment]
@Eugeny Did you (or anyone else) have any success using pug with Angular 20 or 21?

**Comment by raykin** [comment]
@mihilion  If you only use the basic feature of pug, you can try https://github.com/raykin/puglite. I've used it in Angular 20+, should work for Angular 18+

**Comment by Eugeny** [comment]
@mihilion no, we've bit the bullet and converted everything to HTML in the end

**Comment by wildhart** [comment]
@mihilion not yet. Stuck on Angular v19 hoping someone posts a solution here. Got a huge project with too many pug templates to convert to HTML.

**Comment by alan-agius4** [comment]
Hi everyone, thanks for the discussion and for sharing your use cases.

We have decided not to support arbitrary template pre-processing hooks in the application builder (`@angular/build`). 

Here is why:

1. **Language Service & Template Type-Checking:**
   Angular's AOT compiler (`ngtsc`), template type-checker (TCB), and the Angular Language Service rely on accurate character-level source mappings to the original template file. Injecting an external transformation (such as Pug to HTML) breaks source mapping, causing template type-checking errors, jump-to-definition, IDE diagnostics, and refactoring tools to point to intermediate generated HTML rather than the author's source code.

2. **Template Architecture & Modern Control Flow:**
   Angular's modern template pipeline is deeply integrated with the Angular HTML template specification, including native control flow (`@if`, `@for`, `@switch`), `@defer`, and `@let`. Third-party template preprocessors introduce conflicting syntax and make maintaining hydration and compiler invariants unreliable.

3. **HTML is the Official Template Format:**
   HTML is the standard and only officially supported template format for Angular. While Webpack previously allowed third-party loaders to hook into the asset pipeline, that was an artifact of Webpack's architecture rather than an intentional feature of the Angular compiler.

For teams maintaining existing Pug templates, the recommended path forward is to either migrate templates to standard HTML or use an external pre-build step (e.g. a watcher script that compiles `.pug` files into `.html` files before the Angular build processes them).

We understand this may be disappointing for teams with large legacy Pug codebases, but keeping the compiler and language service experience robust and performant requires standardizing on HTML. 

Thank you for your understanding!
